In [1]:
import numpy as np
import pandas as pd

from linearmodels import PooledOLS          # Pooled model
from linearmodels import RandomEffects      # Random-effect model
from linearmodels import PanelOLS           # Fixed-effect model
from linearmodels import FirstDifferenceOLS # First difference model

from linearmodels.panel import compare      # Compare the results of multiple models
from statsmodels.api import add_constant    # for matrices of regression design

In [2]:
df= pd.read_csv('Grunfeld.csv')
df.head()

firm  year    inv   value  capital
0     1  1935  317.6  3078.5      2.8
1     1  1936  391.8  4661.7     52.6
2     1  1937  410.6  5387.1    156.9
3     1  1938  257.7  2792.2    209.2
4     1  1939  330.8  4313.2    203.4

In [3]:
panel_df=df.set_index(['firm','year'])
panel_df.head()

inv   value  capital
firm year                        
1    1935  317.6  3078.5      2.8
     1936  391.8  4661.7     52.6
     1937  410.6  5387.1    156.9
     1938  257.7  2792.2    209.2
     1939  330.8  4313.2    203.4

In [ ]:
mod_pl=PooledOLS.from_formula(formula='inv~1+value+capital',data=panel_df)
res_pl=mod_pl.fit()
res_pl

PooledOLS Estimation Summary                          
================================================================================
Dep. Variable:                    inv   R-squared:                        0.8124
Estimator:                  PooledOLS   R-squared (Between):              0.8357
No. Observations:                 200   R-squared (Within):               0.7385
Date:                Fri, Oct 04 2024   R-squared (Overall):              0.8124
Time:                        19:17:31   Log-likelihood                   -1191.8
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      426.58
Entities:                          10   P-value                           0.0000
Avg Obs:                       20.000   Distribution:                   F(2,197)
Min Obs:                       20.000                                           
Max Obs:                       20.000   F-statistic (robust):             426.58
                                        P-value                           0.0000
Time periods:                      20   Distribution:                   F(2,197)
Avg Obs:                      10.0000                                           
Min Obs:                      10.0000                                           
Max Obs:                      10.0000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept     -42.714     9.5117    -4.4907     0.0000     -61.472     -23.957
value          0.1156     0.0058     19.803     0.0000      0.1041      0.1271
capital        0.2307     0.0255     9.0548     0.0000      0.1804      0.2809
==============================================================================
PanelResults, id: 0x1b17ea9b910

In [ ]:
mod_re = RandomEffects.from_formula(formula='inv~1+value+capital', data=panel_df)
res_re = mod_re.fit()
res_re

RandomEffects Estimation Summary                        
================================================================================
Dep. Variable:                    inv   R-squared:                        0.7695
Estimator:              RandomEffects   R-squared (Between):              0.8148
No. Observations:                 200   R-squared (Within):               0.7667
Date:                Fri, Oct 04 2024   R-squared (Overall):              0.8033
Time:                        19:20:08   Log-likelihood                   -1075.5
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      328.84
Entities:                          10   P-value                           0.0000
Avg Obs:                       20.000   Distribution:                   F(2,197)
Min Obs:                       20.000                                           
Max Obs:                       20.000   F-statistic (robust):             328.84
                                        P-value                           0.0000
Time periods:                      20   Distribution:                   F(2,197)
Avg Obs:                      10.0000                                           
Min Obs:                      10.0000                                           
Max Obs:                      10.0000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept     -57.834     28.899    -2.0013     0.0467     -114.83     -0.8434
value          0.1098     0.0105     10.463     0.0000      0.0891      0.1305
capital        0.3081     0.0172     17.934     0.0000      0.2742      0.3420
==============================================================================
RandomEffectsResults, id: 0x1b1026f94f0

In [ ]:
mod_fe = PanelOLS.from_formula(formula='inv~1+value+capital', data=panel_df)
res_fe = mod_fe.fit()
res_fe

PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:                    inv   R-squared:                        0.8124
Estimator:                   PanelOLS   R-squared (Between):              0.8357
No. Observations:                 200   R-squared (Within):               0.7385
Date:                Fri, Oct 04 2024   R-squared (Overall):              0.8124
Time:                        19:24:34   Log-likelihood                   -1191.8
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      426.58
Entities:                          10   P-value                           0.0000
Avg Obs:                       20.000   Distribution:                   F(2,197)
Min Obs:                       20.000                                           
Max Obs:                       20.000   F-statistic (robust):             426.58
                                        P-value                           0.0000
Time periods:                      20   Distribution:                   F(2,197)
Avg Obs:                      10.0000                                           
Min Obs:                      10.0000                                           
Max Obs:                      10.0000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept     -42.714     9.5117    -4.4907     0.0000     -61.472     -23.957
value          0.1156     0.0058     19.803     0.0000      0.1041      0.1271
capital        0.2307     0.0255     9.0548     0.0000      0.1804      0.2809
==============================================================================


PanelEffectsResults, id: 0x1b102719d90

In [ ]:
dependent = ['inv']
regressors = ['value', 'capital']

y = panel_df[dependent+regressors].dropna()[dependent]
X = add_constant( panel_df[dependent+regressors].dropna()[regressors] )
# For FD-estimator we do not include intercept
X_fd = panel_df[dependent+regressors].dropna()[regressors]

In [ ]:
mod_pl = PooledOLS(y, X)
mod_re = RandomEffects(y, X)
mod_fe = PanelOLS(y, X, entity_effects=True, drop_absorbed=True)


res_pl = mod_pl.fit()
res_re = mod_re.fit()
res_fe = mod_fe.fit()


In [ ]:
compare({'Pool': res_pl, 'RE': res_re, 'FE': res_fe}, stars=True, precision='std_errors')

Model Comparison                           
=======================================================================
                                  Pool                RE             FE
-----------------------------------------------------------------------
Dep. Variable                      inv               inv            inv
Estimator                    PooledOLS     RandomEffects       PanelOLS
No. Observations                   200               200            200
Cov. Est.                   Unadjusted        Unadjusted     Unadjusted
R-squared                       0.8124            0.7695         0.7668
R-Squared (Within)              0.7385            0.7667         0.7668
R-Squared (Between)             0.8357            0.8148         0.8141
R-Squared (Overall)             0.8124            0.8033         0.8027
F-statistic                     426.58            328.84         309.01
P-value (F-stat)                0.0000            0.0000         0.0000
=====================     ============   ===============   ============
const                       -42.714***         -57.834**     -58.744***
                              (9.5117)          (28.899)       (12.454)
value                        0.1156***         0.1098***      0.1101***
                              (0.0058)          (0.0105)       (0.0119)
capital                      0.2307***         0.3081***      0.3101***
                              (0.0255)          (0.0172)       (0.0174)
======================= ============== ================= ==============
Effects                                                          Entity
-----------------------------------------------------------------------

Std. Errors reported in parentheses
PanelModelComparison, id: 0x1b1027098e0

In [ ]:
mod_pl = PooledOLS.from_formula(formula='inv~1+value+capital', data=panel_df)
mod_re = RandomEffects.from_formula(formula='inv~1+value+capital', data=panel_df)
mod_fe = PanelOLS.from_formula(formula='inv~1+value+capital+EntityEffects', data=panel_df)

res_pl = mod_pl.fit(cov_type='clustered', cluster_entity=True)
res_re = mod_re.fit(cov_type='clustered', cluster_entity=True)
res_fe = mod_fe.fit(cov_type='clustered', cluster_entity=True)

compare({'Pool': res_pl, 'RE': res_re, 'FE': res_fe}, stars=True)

Model Comparison                          
=====================================================================
                                 Pool                RE            FE
---------------------------------------------------------------------
Dep. Variable                     inv               inv           inv
Estimator                   PooledOLS     RandomEffects      PanelOLS
No. Observations                  200               200           200
Cov. Est.                   Clustered         Clustered     Clustered
R-squared                      0.8124            0.7695        0.7668
R-Squared (Within)             0.7385            0.7667        0.7668
R-Squared (Between)            0.8357            0.8148        0.8141
R-Squared (Overall)            0.8124            0.8033        0.8027
F-statistic                    426.58            328.84        309.01
P-value (F-stat)               0.0000            0.0000        0.0000
=====================     ===========   ===============   ===========
Intercept                   -42.714**         -57.834**     -58.744**
                            (-2.1989)         (-2.4478)     (-2.2377)
value                       0.1156***         0.1098***     0.1101***
                             (7.6448)          (8.3914)      (7.6205)
capital                     0.2307***         0.3081***     0.3101***
                             (2.8546)          (5.8932)      (6.1803)
======================= ============= ================= =============
Effects                                                        Entity
---------------------------------------------------------------------

T-stats reported in parentheses
PanelModelComparison, id: 0x1b102701e20

In [ ]:

(res_fe.params-res_re.params).T@np.linalg.inv(res_fe.cov-res_re.cov)@(res_fe.params-res_re.params)

-0.0012097695281478617

In [ ]:
from scipy.stats import chi2
chi2.ppf(q=1-0.05, df=res_re.df_model)

7.814727903251179

The test statistic -0.0012 $<$ that the critical point 7.8147 , we do not reject h0  RE is better than FE

In [ ]:
df= pd.read_csv('Guns.csv')
df.head()

year  violent  murder  robbery  prisoners      afam      cauc      male  \
0  1977    414.4    14.2     96.8         83  8.384873  55.12291  18.17441   
1  1978    419.1    13.3     99.1         94  8.352101  55.14367  17.99408   
2  1979    413.3    13.2    109.5        144  8.329575  55.13586  17.83934   
3  1980    448.5    13.2    132.1        141  8.408386  54.91259  17.73420   
4  1981    470.5    11.9    126.5        149  8.483435  54.92513  17.67372   

   population    income   density    state law  
0    3.780403  9563.148  0.074552  Alabama  no  
1    3.831838  9932.000  0.075567  Alabama  no  
2    3.866248  9877.028  0.076245  Alabama  no  
3    3.900368  9541.428  0.076829  Alabama  no  
4    3.918531  9548.351  0.077187  Alabama  no

In [ ]:
panel_df=df.set_index(['state','year'])
panel_df.head()

violent  murder  robbery  prisoners      afam      cauc  \
state   year                                                            
Alabama 1977    414.4    14.2     96.8         83  8.384873  55.12291   
        1978    419.1    13.3     99.1         94  8.352101  55.14367   
        1979    413.3    13.2    109.5        144  8.329575  55.13586   
        1980    448.5    13.2    132.1        141  8.408386  54.91259   
        1981    470.5    11.9    126.5        149  8.483435  54.92513   

                  male  population    income   density law  
state   year                                                
Alabama 1977  18.17441    3.780403  9563.148  0.074552  no  
        1978  17.99408    3.831838  9932.000  0.075567  no  
        1979  17.83934    3.866248  9877.028  0.076245  no  
        1980  17.73420    3.900368  9541.428  0.076829  no  
        1981  17.67372    3.918531  9548.351  0.077187  no

In [ ]:
mod_pl=PooledOLS.from_formula(formula='np.log(violent)~1+law+income+density',data=panel_df)
res_pl=mod_pl.fit()
res_pl

PooledOLS Estimation Summary                          
================================================================================
Dep. Variable:        np.log(violent)   R-squared:                        0.2736
Estimator:                  PooledOLS   R-squared (Between):              0.3256
No. Observations:                1173   R-squared (Within):              -0.3647
Date:                Fri, Oct 04 2024   R-squared (Overall):              0.2736
Time:                        21:03:32   Log-likelihood                   -963.30
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      146.80
Entities:                          51   P-value                           0.0000
Avg Obs:                       23.000   Distribution:                  F(3,1169)
Min Obs:                       23.000                                           
Max Obs:                       23.000   F-statistic (robust):             146.80
                                        P-value                           0.0000
Time periods:                      23   Distribution:                  F(3,1169)
Avg Obs:                       51.000                                           
Min Obs:                       51.000                                           
Max Obs:                       51.000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      5.1236     0.0922     55.544     0.0000      4.9426      5.3046
law[T.yes]    -0.4002     0.0378    -10.592     0.0000     -0.4743     -0.3261
income      6.985e-05  6.714e-06     10.403     0.0000   5.667e-05   8.302e-05
density        0.1201     0.0127     9.4339     0.0000      0.0951      0.1451
==============================================================================
PanelResults, id: 0x1b102807760

In [ ]:
from math import exp
print ('When it allowed to have guns the violence rate decreases by',(exp(-0.40)-1)*100,'%')

When it allowed to have guns the violence rate decreases by -32.967995396436066 %


In [ ]:
mod_re = RandomEffects.from_formula(formula='np.log(violent)~1+law+income+density', data=panel_df)
res_re = mod_re.fit()
res_re

RandomEffects Estimation Summary                        
================================================================================
Dep. Variable:        np.log(violent)   R-squared:                        0.1195
Estimator:              RandomEffects   R-squared (Between):              0.1511
No. Observations:                1173   R-squared (Within):               0.1176
Date:                Fri, Oct 04 2024   R-squared (Overall):              0.1486
Time:                        21:03:33   Log-likelihood                    407.56
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      52.871
Entities:                          51   P-value                           0.0000
Avg Obs:                       23.000   Distribution:                  F(3,1169)
Min Obs:                       23.000                                           
Max Obs:                       23.000   F-statistic (robust):             52.871
                                        P-value                           0.0000
Time periods:                      23   Distribution:                  F(3,1169)
Avg Obs:                       51.000                                           
Min Obs:                       51.000                                           
Max Obs:                       51.000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      5.4783     0.0881     62.213     0.0000      5.3055      5.6511
law[T.yes]     0.0324     0.0178     1.8194     0.0691     -0.0025      0.0672
income       3.76e-05  3.534e-06     10.639     0.0000   3.067e-05   4.454e-05
density        0.0712     0.0373     1.9091     0.0565     -0.0020      0.1444
==============================================================================
RandomEffectsResults, id: 0x1b10276fdf0

In [ ]:
mod_fe = PanelOLS.from_formula(formula='np.log(violent)~1+law+income+density+EntityEffects', data=panel_df)
res_fe = mod_fe.fit()
res_fe

PanelOLS Estimation Summary                           
================================================================================
Dep. Variable:        np.log(violent)   R-squared:                        0.1198
Estimator:                   PanelOLS   R-squared (Between):              0.0407
No. Observations:                1173   R-squared (Within):               0.1198
Date:                Fri, Oct 04 2024   R-squared (Overall):              0.0467
Time:                        21:03:33   Log-likelihood                    440.98
Cov. Estimator:            Unadjusted                                           
                                        F-statistic:                      50.779
Entities:                          51   P-value                           0.0000
Avg Obs:                       23.000   Distribution:                  F(3,1119)
Min Obs:                       23.000                                           
Max Obs:                       23.000   F-statistic (robust):             50.779
                                        P-value                           0.0000
Time periods:                      23   Distribution:                  F(3,1119)
Avg Obs:                       51.000                                           
Min Obs:                       51.000                                           
Max Obs:                       51.000                                           
                                                                                
                             Parameter Estimates                              
==============================================================================
            Parameter  Std. Err.     T-stat    P-value    Lower CI    Upper CI
------------------------------------------------------------------------------
Intercept      5.5332     0.0537     103.05     0.0000      5.4279      5.6386
law[T.yes]     0.0414     0.0178     2.3247     0.0203      0.0065      0.0763
income       3.56e-05  3.569e-06     9.9738     0.0000    2.86e-05    4.26e-05
density       -0.0129     0.0519    -0.2495     0.8031     -0.1147      0.0888
==============================================================================

F-test for Poolability: 222.92
P-value: 0.0000
Distribution: F(50,1119)

Included effects: Entity
PanelEffectsResults, id: 0x1b102a2a1c0

In [ ]:
print ('When it allowed to have guns the violence rate increases by',(exp(0.041)-1)*100,'%')

When it allowed to have guns the violence rate increases by 4.185210554547947 %


significant(because p-value<0.05)

In [ ]:
dependent=['violent']
regressors=['law','income','density']

y=np.log(panel_df[dependent+regressors].dropna()[dependent])
X=add_constant(panel_df[dependent+regressors].dropna()[dependent])

In [ ]:
mod_pl = PooledOLS(y, X)
mod_re = RandomEffects(y, X)
mod_fe = PanelOLS(y, X, entity_effects=True, drop_absorbed=True)


res_pl = mod_pl.fit()
res_re = mod_re.fit()
res_fe = mod_fe.fit()

In [ ]:
compare({'Pool': res_pl, 'RE': res_re, 'FE': res_fe}, stars=True, precision='std_errors')

Model Comparison                            
=========================================================================
                                   Pool                RE              FE
-------------------------------------------------------------------------
Dep. Variable                   violent           violent         violent
Estimator                     PooledOLS     RandomEffects        PanelOLS
No. Observations                   1173              1173            1173
Cov. Est.                    Unadjusted        Unadjusted      Unadjusted
R-squared                        0.7727            0.6615          0.6555
R-Squared (Within)               0.6063            0.6554          0.6555
R-Squared (Between)              0.7862            0.7476          0.7436
R-Squared (Overall)              0.7727            0.7406          0.7369
F-statistic                      3980.5            2287.9          2133.1
P-value (F-stat)                 0.0000            0.0000          0.0000
=====================     =============   ===============   =============
const                         5.1731***         5.3471***       5.3568***
                               (0.0163)          (0.0435)        (0.0148)
violent                       0.0017***         0.0014***       0.0013***
                            (2.691e-05)       (2.827e-05)     (2.886e-05)
======================= =============== ================= ===============
Effects                                                            Entity
-------------------------------------------------------------------------

Std. Errors reported in parentheses
PanelModelComparison, id: 0x1b102a2b4c0

In [ ]:
mod_pl = PooledOLS.from_formula(formula='np.log(violent)~1+law+income+density', data=panel_df)
mod_re = RandomEffects.from_formula(formula='np.log(violent)~1+law+income+density', data=panel_df)
mod_fe = PanelOLS.from_formula(formula='np.log(violent)~1+law+income+density+EntityEffects', data=panel_df)

res_pl = mod_pl.fit(cov_type='clustered', cluster_entity=True)
res_re = mod_re.fit(cov_type='clustered', cluster_entity=True)
res_fe = mod_fe.fit(cov_type='clustered', cluster_entity=True)

compare({'Pool': res_pl, 'RE': res_re, 'FE': res_fe}, stars=True)

Model Comparison                                 
===================================================================================
                                       Pool                  RE                  FE
-----------------------------------------------------------------------------------
Dep. Variable               np.log(violent)     np.log(violent)     np.log(violent)
Estimator                         PooledOLS       RandomEffects            PanelOLS
No. Observations                       1173                1173                1173
Cov. Est.                         Clustered           Clustered           Clustered
R-squared                            0.2736              0.1195              0.1198
R-Squared (Within)                  -0.3647              0.1176              0.1198
R-Squared (Between)                  0.3256              0.1511              0.0407
R-Squared (Overall)                  0.2736              0.1486              0.0467
F-statistic                          146.80              52.871              50.779
P-value (F-stat)                     0.0000              0.0000              0.0000
=====================     =================   =================   =================
Intercept                         5.1236***           5.4783***           5.5332***
                                   (16.990)            (38.584)            (40.575)
law[T.yes]                       -0.4002***              0.0324              0.0414
                                  (-2.7652)            (0.8354)            (1.0527)
income                         6.985e-05***         3.76e-05***         3.56e-05***
                                   (3.2962)            (4.1013)            (3.7701)
density                           0.1201***           0.0712***             -0.0129
                                   (9.3320)            (5.1328)           (-0.4173)
======================= =================== =================== ===================
Effects                                                                      Entity
-----------------------------------------------------------------------------------

T-stats reported in parentheses
PanelModelComparison, id: 0x1b1028070a0

both in random and in panel are not significant

In [ ]:
(res_fe.params-res_re.params).T@np.linalg.inv(res_fe.cov-res_re.cov)@(res_fe.params-res_re.params)

2.1770131320687147

In [ ]:
chi2.ppf(q=1-0.05, df=res_re.df_model)

9.487729036781154

The test statistic is lower than the critical point $\Rightarrow$ we do not reject the $H_0$ $\Rightarrow$ RE is better than FE

In [ ]:
df= pd.read_csv('EmplUK.csv')
df.head()

firm  year  sector    emp     wage  capital      output
0     1  1977       7  5.041  13.1516   0.5894   95.707199
1     1  1978       7  5.600  12.3018   0.6318   97.356903
2     1  1979       7  5.015  12.8395   0.6771   99.608299
3     1  1980       7  4.715  13.8039   0.6171  100.550100
4     1  1981       7  4.093  14.2897   0.5076   99.558098

In [ ]:
panel_df=df.set_index(['firm','year'])
panel_df.head()

sector    emp     wage  capital      output
firm year                                             
1    1977       7  5.041  13.1516   0.5894   95.707199
     1978       7  5.600  12.3018   0.6318   97.356903
     1979       7  5.015  12.8395   0.6771   99.608299
     1980       7  4.715  13.8039   0.6171  100.550100
     1981       7  4.093  14.2897   0.5076   99.558098

In [ ]:
mod_pl = PooledOLS.from_formula(formula='np.log(emp)~1+np.log(wage)+np.log(capital)+np.log(output)', data=panel_df)
mod_re = RandomEffects.from_formula(formula='np.log(emp)~1+np.log(wage)+np.log(capital)+np.log(output)', data=panel_df)
mod_fe = PanelOLS.from_formula(formula='np.log(emp)~1+np.log(wage)+np.log(capital)+np.log(output)+EntityEffects', data=panel_df)

res_pl = mod_pl.fit(cov_type='clustered', cluster_entity=True)
res_re = mod_re.fit(cov_type='clustered', cluster_entity=True)
res_fe = mod_fe.fit(cov_type='clustered', cluster_entity=True)

compare({'Pool': res_pl, 'RE': res_re, 'FE': res_fe}, stars=True)

Model Comparison                            
=========================================================================
                                   Pool                RE              FE
-------------------------------------------------------------------------
Dep. Variable               np.log(emp)       np.log(emp)     np.log(emp)
Estimator                     PooledOLS     RandomEffects        PanelOLS
No. Observations                   1031              1031            1031
Cov. Est.                     Clustered         Clustered       Clustered
R-squared                        0.8356            0.6646          0.6143
R-Squared (Within)               0.5340            0.6064          0.6143
R-Squared (Between)              0.8476            0.8071          0.7559
R-Squared (Overall)              0.8356            0.7988          0.7499
F-statistic                      1740.1            678.24          471.39
P-value (F-stat)                 0.0000            0.0000          0.0000
=====================     =============   ===============   =============
Intercept                        0.3444            0.2237         -0.2159
                               (0.2713)          (0.3727)       (-0.3550)
np.log(wage)                   -0.3669*        -0.2900***      -0.3106***
                              (-1.7191)         (-2.6588)       (-2.7097)
np.log(capital)               0.8090***         0.6392***       0.5489***
                               (24.796)          (18.773)        (11.254)
np.log(output)                 0.4791**         0.4401***       0.5370***
                               (2.3939)          (4.6266)        (5.2730)
======================= =============== ================= ===============
Effects                                                            Entity
-------------------------------------------------------------------------

T-stats reported in parentheses
PanelModelComparison, id: 0x1b102a8b2b0

In [ ]:
(res_fe.params-res_re.params).T@np.linalg.inv(res_fe.cov-res_re.cov)@(res_fe.params-res_re.params)

27.723280421136835

In [ ]:
chi2.ppf(q=1-0.05, df=res_re.df_model)

9.487729036781154

The test statistic is higher than the critical point $\Rightarrow$ we reject the $H_0$ $\Rightarrow$ FE is better than RE